pocs/camera/libasi.py

```diff
@@ -0,0 +1,597 @@
+                         'supported_bins': self._parse_bins(camera_info.supported_bins),
+                         'supported_video_format': self._parse_formats(
+                             camera_info.supported_video_format),
+                         'pixel_size': camera_info.pixel_size * u.um,
+                         'has_mechanical_shutter': bool(camera_info.has_mechanical_shutter),
+                         'has_ST4_port': bool(camera_info.has_ST4_port),
+                         'has_cooler': bool(camera_info.has_cooler),
+                         'is_USB3_host': bool(camera_info.is_USB3_host),
+                         'is_USB3_camera': bool(camera_info.is_USB3_camera),
+                         'e_per_adu': camera_info.e_per_adu * u.electron / u.adu,
+                         'bit_depth': camera_info.bit_depth * u.bit,
+                         'is_trigger_camera': bool(camera_info.is_trigger_camera)}
+        return pythonic_info
+
+    def _parse_bins(self, supported_bins):
+        bins = tuple(int(bin) for bin in supported_bins if bin != 0)
+        return bins
+
+    def _parse_formats(self, supported_formats):
+        formats = []
+        for i in range(8):
+            format = ImgType(supported_formats[i])
+            if format != ImgType.END:
+                formats.append(format.name)
+            else:
+                break
+        return tuple(formats)
+
+    def _parse_caps(self, control_caps):
+        """ Utility function to parse ControlCaps Structures into something more Pythonic """
+        control_type = ControlType(control_caps.control_type).name
+        control_info = {'name': control_caps.name.decode(),
+                        'description': control_caps.description.decode(),
+                        'max_value': self._parse_return_value(control_caps.max_value,
+                                                              control_type),
+                        'min_value': self._parse_return_value(control_caps.min_value,
+                                                              control_type),
+                        'default_value': self._parse_return_value(control_caps.default_value,
+                                                                  control_type),
+                        'is_auto_supported': bool(control_caps.is_auto_supported),
+                        'is_writable': bool(control_caps.is_writable),
+                        'control_type': control_type}
+        return control_info
+
+    def _parse_return_value(self, value, control_type):
+        """ Helper function to apply appropiate type conversion and/or units to value """
+        try:
+            int_value = value.value  # If not done already extract Python int from ctypes.c_long
+        except AttributeError:
+            int_value = value  # If from a ctypes struct value will already be a Python int
+
+        # Apply control type specific units and/or data types
+        if control_type == 'EXPOSURE':
```

Comment:
Hmmm. Agree it's ugly, but not obvious to me how to do it more elegantly, because of the mix of casting, scale factors, unit conversions, enum lookup, etc.  A dictionary of lambda functions would do it, but wouldn't really simplify things. I could group all the bools together in one `elif control_type in (<list of control types>): nice_value = bool(int_value)` though, which might help a bit.

Reply:
The order follows the order of the control types in the `enum` that lists them all. Now you raise it I'm not sure there's any reason to stick with that though.

Reply:
Actually there are only 4 distinct cases, so with the help of a dictionary and a tuple I've reduces it to a 4 block if, elif, elif, else. Looks nicer now.